### ISSUE TITLE
Proposal to Revise Section 4.4 to Support ISO 8601 in Time Coordinate Representation

### ISSUE THREAD

**Issue by cofinoa** [cf-convention/cf-conventions#581]
# Title
Proposal to Revise Section 4.4 to Support ISO 8601 and Enhance Clarity in Time Coordinate Representation

# Moderator
TBC

# Moderator Status Review [last updated: 2024-12-22]
Initial submission. Awaiting community feedback and discussion.

# Requirement Summary
This proposal seeks to enhance Section 4.4 of the CF Metadata Conventions by introducing explicit support for ISO 8601 string representations of time coordinates. The update aims to improve clarity, modernize the conventions for interoperability with external systems, and maintain backward compatibility with existing datasets.

# Technical Proposal Summary
- Add ISO 8601-based string representations (`YYYY-MM-DD`, `YYYY-MM-DDTHH:MM:SSZ`, `YYYYMMDDTHHMMSSZ`) as a valid alternative to relative-based representations for time coordinates.
- Provide examples for both relative-based and string-based time coordinates.
- Preserve backward compatibility with existing CF-compliant datasets.

# Benefits
 - **Interoperability**: Ensures compatibility with modern data standards like JSON, REST APIs, and ISO-compliant systems.
 - **Readability**: Enhances dataset usability by allowing self-descriptive, human-readable time strings.
 - **Sorting**: ISO 8601 strings are naturally lexicographically sortable, improving data management workflows.
 - **Legalization of Previously Invalid Representations**: Examples in Section 7.4 of the CF Conventions, which use textual time representations for climatologies, currently describe formats that are not permitted by the standard. This proposal will make such examples valid and consistent with the conventions.
 - **Backward Compatibility**: Maintains support for existing relative-based time coordinate representations.

# Status Quo
Currently, Section 4.4 of the CF Metadata Conventions only supports numeric offsets from a reference datetime (e.g., `days since 2000-1-1`). While effective, this approach lacks the direct compatibility with ISO 8601 datetime strings that is commonly used in APIs, REST systems, and JSON-based formats.

# Associated pull request
[To be linked once submitted]

# Detailed Proposal
## 1. Overview
This revision introduces support for ISO 8601 datetime representations alongside existing relative-based formats. It ensures clarity and flexibility while maintaining backward compatibility with legacy datasets.

## 2. Proposed Changes
 - **Support ISO 8601 Formats in `units`**:
    Add the ability to specify ISO 8601 datetime strings directly, with examples such as:
    - `units = "ISO 8601:YYYY-MM-DD";`
    - `units = "ISO 8601:YYYY-MM-DDTHH:MM:SSZ";`
    - `units = "ISO 8601:YYYYMMDDTHHMMSSZ"; `
   
 - **Provide Comprehensive Examples**:
   - Example: Relative-based Representation
      ```
      double time(time) ;  
        time:axis = "T" ;  
        time:units = "days since 1990-01-01 00:00:00" ;  

        // Example data:  
       data:  
         time = 0.0, 1.0, 1.5, 5.0 ;  
      ```

   - Example: String-based Representation
      ```
      string time(time) ;  
         time:axis = "T" ;  
         time:units = "ISO 8601:YYYY-MM-DDTHH:MM:SSZ" ;  

        // Example data:  
       data:  
         time = "1990-01-01T00:00:00Z", "1990-01-02T00:00:00Z", "1990-01-02T12:00:00Z", "1990-01-06T00:00:00Z" ;  
      ```
## 3. Backward Compatibility
The proposal maintains full backward compatibility. Datasets using relative-based representations (`days since <datetime>`) remain unaffected.

**Comment by cofinoa** [comment]
Updated *2. Propossed Changes* replacing the optional `axis` attribute with the optional `standard_name` attribute

**Comment by JonathanGregory** [comment]
Dear Antonio @cofinoa

Thanks for making this careful proposal, to add support in CF for ISO8601 datetime string-valued coordinates as an alternative to numeric coordinates. Although I recognise there would be benefits of convenience in some circumstances from making this change, it wouldn't enable anything which is currently not possible with CF. Therefore I think the costs and difficulties are greater than the advantages, and I don't support the change.

Because it's an attractive idea, it has been proposed before, and decided against, at least twice during the development of CF. It was proposed and debated in 2007 in [trac ticket 14](https://cfconventions.org/Data/Trac-tickets/14.html). It was suggested and discussed again in a thread that concluded with the subject "time as ISO strings" on the CF email list in 2010 e.g. [Steve Hankin's email](https://cfconventions.org/mailing-list-archive/Data/3730.html).

I don't think we should make this change "because all previous versions must generally continue to be supported in software for the sake of archived datasets, and in order to limit the complexity of the conventions. For these reasons, there is a strong preference against introducing any new capability to the conventions when there is already some method that can adequately serve the same purpose (even if a different method would arguably be better than the existing one)" (principle 10 in Sect 1.2).

In the previous discussions, several specific points have been made, including:

* ISO datetime strings are not defined for non-real-world calendars.

* ISO datetime strings raise issues of precision, because omitted elements imply imprecise datetimes, which is a concept CF doesn't have at present.

* You can't work out time-intervals from datetime strings without converting them into numbers.

* Datetime strings take up more storage than double-precision numbers.

* Although numeric time coordinate values themselves aren't human-readable datetimes, software is available to convert them into strings&mdash;probably more software than when we last discussed this issue. In particular, `ncdump -t` converts CF-netCDF numeric time coordinates to human-readable strings (for all CF calendars except for the leap-second-aware ones we've added in CF1.12).

Best wishes

Jonathan

**Comment by cofinoa** [comment]
Dear @JonathanGregory,

Thank you for your thoughtful and detailed reply. I appreciate you taking the time to revisit this proposal and providing the historical context and references to past discussions, as well as highlighting the implications of Principle 10 in Section 1.2 of the CF Conventions.

One of my key concerns with the **current numeric representation** is that it transforms an **absolute datetime** (e.g., `1990-01-01T00:00:00Z`) into a **relative period** measured from a reference epoch. While effective for many use cases, this approach introduces challenges when integrating CF-compliant data with modern systems, APIs, and formats that rely on absolute datetime strings for interoperability. This transformation adds a layer of abstraction that can lead to potential pitfalls, particularly when human readability and self-descriptive formats are priorities.

Additionally, the numeric representation assumes regular time intervals, which works well for most calendars but introduces complications with **UTC**, as described in Section 4.4.3 of the CF Conventions. UTC includes **leap seconds**, adjustments made to account for irregularities. These adjustments mean that intervals between times are not always uniform, and a single numeric time value cannot directly represent a leap second (e.g., `23:59:60`) without additional metadata. This makes conversions between numeric representations and calendar datetimes more complex and potentially error-prone, particularly for datasets requiring precise temporal alignment.

I also understand that using **ISO 8601** for string-based time representation may implicitly tie it to the Gregorian calendar, potentially introducing confusion when applied to non-real-world calendars like `noleap` or `360_day`. My intention is not to enforce ISO 8601’s calendar semantics but to leverage its **string format templates** to represent date and time values in these calendars. If this approach creates confusion or misalignment with the standard, I would be open to exploring alternative naming for the representation, rather than strictly calling it "ISO 8601" (i.e. CF-DATETIME).

I understand your concerns regarding **Principle 10**, which emphasizes the importance of limiting new capabilities when existing methods already address the same needs. While numeric representations can adequately serve many use cases, I believe that introducing string-based formats as an optional alternative addresses significant gaps in modern interoperability without disrupting existing practices.

This proposal helps to improve CF while aligning with several principles of the CF Conventions:

- **Principle 1: Self-Description**  
  String-based formats naturally support self-description by encoding complete datetime information directly within the time coordinate, eliminating the need for external tools or additional metadata for interpretation.

- **Principle 2: Addressing Clear Needs**  
  This proposal addresses a clear need for modern interoperability while ensuring that it remains optional and minimally disruptive to existing practices. By leveraging widely used string-based formats (like ISO 8601), it provides a practical solution for emerging use cases.

- **Principle 4: Practicality**  
  By considering string-based formats, the proposal ensures practicality for both data producers and users. Producers benefit from a widely accepted standard, while users gain access to intuitive, human-readable time representations.

- **Principle 5: Human and Machine Readability**  
  String-formats, based in ISO 8601 templates, are both human-readable and easily parsable by programs, enhancing CF’s usability in modern workflows.

- **Principle 7: Minimizing Mistakes**  
  Leap second handling within the UTC calendar exemplifies a scenario where numeric representations may lead to errors. String-based representations such as ISO 8601 explicitly support leap second semantics, ensuring precise and unambiguous temporal metadata and reducing the possibility of mistakes.

- **Principle 8: Flexibility and Optionality**  
  This proposal aligns with Principle 8 by providing string-based formats as an optional alternative to numeric representations. It empowers data producers to describe their data in ways that best suit their workflows and interoperability requirements, while maintaining support for existing practices.

With this in mind, I will take your feedback to refine my proposal further. I plan to review the past discussions and decisions you mentioned, including trac ticket 14 (2007) and the CF email thread from 2010, to better understand the concerns raised at the time and how they were addressed.

Thank you again for your thoughtful response. I look forward to continuing this discussion and refining the proposal to best serve CF’s goals and community needs.

**Comment by cofinoa** [comment]
* Comment on Updates made on 2024-12-31*

The issue description and title have been updated to clarify the proposal's intent and scope. The title now emphasizes the support for string-based datetime representations using ISO 8601 templates, and the description provides a detailed breakdown of the benefits, technical proposal, and alignment with Section 4.4 of the CF Metadata Conventions. These changes aim to enhance clarity and precision for community discussion.

**Comment by JonathanGregory** [comment]
Dear Antonio

Thanks for considering my objections. Here are some further comments.

* I acknowledge that for some purposes string representations of datetimes are necessary or convenient. But that, in my view, is not sufficient reason for adding an alternative representation of time to CF. Doing so would require all software that aims to comply with CF to implement a new set of functions, as well as the existing facilities. Until this is completed, data which used the new string convention would not be interpretable by all software i.e. the new convention would break existing programs. Thus it would harm interoperability. "All software" doesn't just mean professionally maintained packages, but includes ad-hoc software written by data-users to process or analyse data. Given the huge importance of time coordinates, I don't think we should consider causing such disruption unless there is a really essential use-case which _cannot_ be dealt with by the existing conventions.

* Since the time coordinate variable includes the `units` and the `calendar` attributes, I think the numeric convention for time coordinates is nearly self-describing, in the sense that you don't need to refer to external resources. I agree that you have to _know_ what the calendars mean, and in the case of `utc` you have to know when the leap-seconds occurred. The former is fairly common knowledge or self-explanatory, but the latter is something most people have to look up. Although, as you say, this knowledge is not needed for writing and reading formatted datetime strings, you can't escape the need for it eventually, which instead occurs when you want to work out the interval of time between two instants. With numeric time coordinates we have the difficulty that an instrument may not know about leap seconds and hence might produce the wrong number to represent the UTC of an observation. In those circumstances the same ignorant instrument will produce an incorrect datetime string instead.

* You're right of course that times expressed as strings are easily human-readable whereas they are not as floating-point numbers. But netCDF is a binary format, which a human can't read in any case. As a human, to read a netCDF file, you have to use a program of some sort. You can convert it to ASCII using `ncdump`, which can translate CF times into udunits-like strings, or you can use software libraries in Python and other languages which know how to convert numeric time coordinates to and from strings in various formats.

Similar things have been said in the previous discussions.

Let's see what others think. Happy `2025-01-01 00:00:00`.

Jonathan

**Comment by DocOtak** [comment]
I agree with @sethmcg and @JonathanGregory that ISO8601 strings should not be included in CF for reasons stated above and in the linked email from [Steve Hankin](https://cfconventions.org/mailing-list-archive/Data/3730.html).

The tooling in python and in things like ncdump with the `-i` or `-t` flag are good to the point that I don't usually think about how the time data are actually serialized in the netCDF file. In python, I basically never deal with the underlying data format, it basically always using some numpy datetime dtype that lets me use ISO strings in my code.

If I was supportive of this idea, I would suggest that the units be a string that is compatible with the C standard libraries [strptime](https://www.man7.org/linux/man-pages/man3/strptime.3.html) for formats, so e.g. `%Y-%m-%d` rather than `YYYY-MM-DD`.

**Comment by taylor13** [comment]
I also agree with the reasoning of Steve Hankin and others that ISO8601 strings should not be included. (Thanks @DocOtak for reminding us of Steve's often quite helpful input.)

**Comment by DocOtak** [comment]
@taylor13 Credit goes to @JonathanGregory for digging that up, it is indeed very insightful.

**Comment by ChrisBarker-NOAA** [comment]
Another note for posterity:

We *finally* have a UTC calendar in CF, which closes the last missing link (IMHO) that supports this proposal. That is, if a data creator starts with a UTC timestamp (that may have a leap second in it, no matter how unlikely), there IS a way to losslessly (and unambiguously) store it now.

In practice, the software hasn't caught up yet, but hopefully it will before too long ...

**Comment by JonathanGregory** [comment]
I am closing this issue with `agreement not to change` since no-one has dissented. Thanks to all for discussion.

**Comment by cofinoa** [comment]
Thanks all for the careful discussion, and @JonathanGregory for the clear summary and closure.

Just for the record, I agree that it is reasonable to close this issue, since there is not enough support to move the proposal forward at this stage.

Looking back, part of my motivation for opening this issue came from a genuine discomfort with the current situation around datetime representations in Section 4.4: not only possible limitations in the normative model, but also inconsistencies in the way datetimes are described and represented across the document. That combination, together with some edge cases such as leap-second-related handling, was what initially made the topic feel worth addressing.

In that sense, I think issue #624 has helped to separate more clearly the editorial and documentation aspects from the broader normative question. With hindsight, that narrower path may have been a better first step.

I still think the underlying topic may be worth revisiting in the future, particularly if clearer shared use cases emerge from the community. Until then, I am happy for this issue to remain as a reference point.
